File: web/src/ask.js
// The model's questions, normalised.
//
// One card can come from two places, and both must render identically:
//
//   - live, from the `ask_user` event on the stream, which carries the question
//     the moment the model asks it — this is the one that blocks the turn;
//   - from a reloaded conversation, where the question is the arguments of a
//     persisted `ask_user` tool call and the answer is its result.
//
// The server persists both (they are the tool call's arguments and its result),
// so a card survives a refresh without a table of its own. This module is the
// single place that knows how to read either, which is what keeps a reloaded
// card from rendering differently from a live one.

/** The tool whose calls are rendered as cards rather than as tool rows. */
export const ASK_TOOL = 'ask_user'

/** Card states. `pending` is the only one that accepts input. */
export const ASK_PENDING = 'pending'
export const ASK_SUBMITTING = 'submitting'
export const ASK_ANSWERED = 'answered'
export const ASK_TIMEOUT = 'timeout'
export const ASK_CANCELLED = 'cancelled'

/** True for the tool calls that become cards. */
export function isAskTool(name) {
  return String(name || '') === ASK_TOOL
}

function asRecord(value) {
  return value && typeof value === 'object' && !Array.isArray(value) ? value : null
}

function asString(value) {
  return typeof value === 'string' ? value : value === null || value === undefined ? '' : String(value)
}

/** Parse a JSON column that may be a string, an object or (when broken) nothing. */
function parseJSON(value) {
  if (value === null || value === undefined || value === '') return null
  if (typeof value === 'object') return value
  if (typeof value !== 'string') return null
  try {
    return JSON.parse(value)
  } catch (err) {
    // A half-written column must never break the conversation.
    return null
  }
}

/** Normalise the offered options; entries without a usable label are dropped. */
function normalizeOptions(raw) {
  const list = Array.isArray(raw) ? raw : []
  const out = []
  const seen = new Set()
  for (const entry of list) {
    const source = asRecord(entry)
    const label = asString(source ? source.label : entry).trim()
    if (!label || seen.has(label)) continue
    seen.add(label)
    out.push({ label, description: asString(source && source.description).trim() })
  }
  return out
}

function normalizeSelected(raw) {
  const list = Array.isArray(raw) ? raw : []
  return list.map((v) => asString(v).trim()).filter(Boolean)
}

/**
 * Build a card from an `ask_user` stream event.
 *
 * Returns null for the outcome shape: that one updates a card that is already on
 * screen (see `applyAskOutcome`), which is why it carries an id and no question.
 */
export function askFromEvent(event) {
  if (!event || typeof event !== 'object') return null
  const ask = asRecord(event.ask)
  if (!ask) return null
  const id = asString(ask.id).trim()
  if (!id) return null
  return {
    id,
    header: asString(ask.header).trim(),
    question: asString(ask.text),
    options: normalizeOptions(ask.options),
    multiSelect: Boolean(ask.multi_select),
    // Absent means "allowed": the server always states it, but a card that lost
    // the field must not silently remove the free-text box the model expected.
    allowCustom: ask.allow_custom === undefined ? true : Boolean(ask.allow_custom),
    status: ASK_PENDING,
    selected: [],
    text: '',
    error: '',
    fromHistory: false,
  }
}

/** Apply the outcome event (answered / timeout / cancelled) to a card. */
export function applyAskOutcome(card, event) {
  if (!card || !event || typeof event !== 'object') return false
  const status = asString(event.ask_status)
  if (!status) return false
  const answer = asRecord(event.ask_answer)
  if (answer) {
    card.selected = normalizeSelected(answer.selected)
    card.text = asString(answer.text)
  }
  // "submitting" is a local state with the request in flight; the server's word
  // is the truth once it arrives.
  card.status = status
  return true
}

/**
 * Build a card from a persisted `ask_user` tool call.
 *
 * `args` is the question and `result` is the outcome, both exactly as the tool
 * saw them. A call that never produced a result was a turn that died mid-ask,
 * and it renders as cancelled — the honest reading of a question the server
 * never got an answer for.
 */
export function askFromTool(tool) {
  const source = asRecord(tool)
  if (!source) return null
  const args = asRecord(parseJSON(source.args))
  if (!args) return null
  const question = asString(args.question)
  if (!question) return null

  const output = asRecord(parseJSON(source.result))
  const id = asString(source.id) || `ask-${asString(source.key)}`
  const card = {
    id,
    header: asString(args.header).trim(),
    question,
    options: normalizeOptions(args.options),
    multiSelect: Boolean(args.multi_select),
    allowCustom: args.allow_custom === undefined ? true : Boolean(args.allow_custom),
    status: ASK_ANSWERED,
    selected: [],
    text: '',
    error: '',
    fromHistory: true,
  }

  if (source.error) {
    card.status = ASK_CANCELLED
    card.error = asString(source.error)
    return card
  }
  if (!output) {
    card.status = ASK_CANCELLED
    return card
  }
  card.status = asString(output.status) || ASK_ANSWERED
  card.selected = normalizeSelected(output.selected)
  card.text = asString(output.text)
  return card
}

/** Every card a persisted or live turn carries, in call order. */
export function askCardsFromTools(tools) {
  const list = Array.isArray(tools) ? tools : []
  const out = []
  for (const tool of list) {
    if (!isAskTool(tool && tool.name)) continue
    const card = askFromTool(tool)
    if (card) out.push(card)
  }
  return out
}

/** True when a card still accepts an answer. */
export function isAskOpen(card) {
  return Boolean(card) && card.status === ASK_PENDING
}

/**
 * One line describing what was answered, for the settled card and for the
 * tests. It mirrors what the server renders into the model's `answer` field, so
 * a reader of the conversation and the model are told the same thing.
 */
export function askAnswerLine(card) {
  const selected = normalizeSelected(card && card.selected)
  const text = asString(card && card.text).trim()
  const joined = selected.join('、')
  if (joined && text) return `${joined}（补充：${text}）`
  return joined || text
}

/** The status label shown on the card's head. */
export function askStatusLabel(card) {
  const status = card && card.status
  if (status === ASK_PENDING) return '等待你的回答'
  if (status === ASK_SUBMITTING) return '提交中…'
  if (status === ASK_ANSWERED) return '已提交'
  if (status === ASK_TIMEOUT) return '已超时'
  if (status === ASK_CANCELLED) return '已取消'
  return ''
}

/** One line explaining a settled card that nobody answered. */
export function askClosedNote(card) {
  const status = card && card.status
  if (status === ASK_TIMEOUT) return '等待超时，模型已按自己的判断继续'
  if (status === ASK_CANCELLED) return '本轮已中断，模型没有等到回答'
  return ''
}

File: internal/server/ask.go
package server

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/cloudwego/hertz/pkg/app"
	"go.uber.org/zap"

	"github.com/huan/huan-agent/internal/chat"
	"github.com/huan/huan-agent/internal/tool"
)

// DefaultAskTimeout bounds how long one of the model's questions waits for an
// answer when the deployment does not say otherwise.
//
// Ten minutes is a compromise: long enough that stepping away from the desk does
// not lose the turn, short enough that a forgotten tab does not pin a goroutine
// and a half-finished tool call for an afternoon. It is bounded again by the
// turn's own context — a stop, a closed tab or a fired deadline ends the wait —
// and by chat.turn_deadline_seconds, which the waiting time counts against.
const DefaultAskTimeout = 10 * time.Minute

// errQuestionGone is what a submission gets when the question it answers is no
// longer waiting: already answered, timed out, cancelled with its turn, or never
// existed. All four are the same 404 on purpose — a client cannot act on the
// difference, and telling a stranger which question ids exist is not worth the
// precision.
var errQuestionGone = errors.New("question is no longer pending")

// pendingQuestion is one question a running turn is waiting on.
type pendingQuestion struct {
	id      string
	session string
	// question is kept so a submission can be checked against what was actually
	// offered: an answer that names an option the card never had is a bug or a
	// forgery, and the model must not be handed either as if a person chose it.
	question tool.Question
	// answers carries the submission to the waiting tool. It is buffered (one
	// slot) so whoever delivers an answer never blocks, even in the moment the
	// waiter has given up.
	answers chan tool.Answer
	askedAt time.Time
}

// questionHub tracks the questions currently waiting for an answer.
//
// It exists because the answer arrives on a different HTTP request from the one
// the turn is streaming on: the run goroutine is parked inside the tool call, and
// this is the only object both requests can see.
type questionHub struct {
	mu      sync.Mutex
	pending map[string]*pendingQuestion
}

func newQuestionHub() *questionHub {
	return &questionHub{pending: map[string]*pendingQuestion{}}
}

// register adds a question and returns its waiting record.
func (h *questionHub) register(session string, q tool.Question) *pendingQuestion {
	p := &pendingQuestion{
		id:       q.ID,
		session:  session,
		question: q,
		answers:  make(chan tool.Answer, 1),
		askedAt:  time.Now(),
	}
	h.mu.Lock()
	h.pending[q.ID] = p
	h.mu.Unlock()
	return p
}

// peek returns the waiting question without removing it, so a submission can be
// validated before it consumes the question.
func (h *questionHub) peek(id, session string) (*pendingQuestion, bool) {
	h.mu.Lock()
	defer h.mu.Unlock()
	p, ok := h.pending[id]
	if !ok || p.session != session {
		return nil, false
	}
	return p, true
}

// resolve hands an answer to the waiting tool, consuming the question. The
// removal and the hand-off happen under one lock, so of two simultaneous
// submissions exactly one wins and the other is told the question is gone.
func (h *questionHub) resolve(id, session string, a tool.Answer) error {
	h.mu.Lock()
	p, ok := h.pending[id]
	if !ok || p.session != session {
		h.mu.Unlock()
		return errQuestionGone
	}
	delete(h.pending, id)
	h.mu.Unlock()

	// Buffered with room for exactly this message: the send cannot block.
	p.answers <- a
	return nil
}

// forget drops a question its waiter has given up on (a timeout or a cancelled
// turn), so a late submission is refused instead of being delivered to nobody.
func (h *questionHub) forget(id string) {
	h.mu.Lock()
	delete(h.pending, id)
	h.mu.Unlock()
}

// count reports how many questions are waiting. It exists for the tests and for
// a leak check: a hub that grows without bound means a waiter stopped forgetting.
func (h *questionHub) count() int {
	h.mu.Lock()
	defer h.mu.Unlock()
	return len(h.pending)
}

// turnAsker is the Asker for one turn: it publishes the question on that turn's
// SSE stream and waits for the answer to arrive through the hub.
type turnAsker struct {
	hub     *questionHub
	session string
	timeout time.Duration
	// emit writes onto the turn's event channel. It is the same channel the
	// runner reports tool calls on, which is what makes the card appear in the
	// conversation at the moment the model asks rather than after it finishes.
	emit   func(chat.Event)
	logger *zap.Logger
}

// Ask publishes a question and waits for the answer.
//
// It never returns an error for "nobody answered": a timeout and a cancelled
// turn are answers with a status, because the model still has to decide what to
// do next. An error here means the machinery failed, which is a different thing
// and must not be reported to the model as if a person had ignored it.
func (a *turnAsker) Ask(ctx context.Context, q tool.Question) (tool.Answer, error) {
	// A Server built without New (tests do that) has no registry to park a
	// question in. Failing here with a sentence beats registering into a nil map
	// and taking the process down from inside a tool call.
	if a.hub == nil {
		return tool.Answer{}, errors.New("server: question registry is not initialized")
	}
	id, err := newQuestionID()
	if err != nil {
		return tool.Answer{}, fmt.Errorf("分配问题 id: %w", err)
	}
	q.ID = id

	pending := a.hub.register(a.session, q)
	// Whichever way this returns, the question stops being answerable.
	defer a.hub.forget(id)

	a.emit(chat.Event{Type: chat.EventAsk, Ask: &q, AskStatus: chat.AskPending})

	var timerC <-chan time.Time
	if a.timeout > 0 {
		timer := time.NewTimer(a.timeout)
		defer timer.Stop()
		timerC = timer.C
	}

	var answer tool.Answer
	select {
	case answer = <-pending.answers:
	case <-timerC:
		answer = tool.Answer{Status: tool.AnswerTimeout}
	case <-ctx.Done():
		// The turn is over: the browser went away, the user pressed 停止, or the
		// process is shutting down. Nobody will answer a question for a turn that
		// no longer exists.
		answer = tool.Answer{Status: tool.AnswerCancelled}
	}

	// Tell the client how it ended. A pending card that is never closed would sit
	// there offering choices for a question nobody is listening to.
	a.emit(chat.Event{
		Type:      chat.EventAsk,
		AskID:     id,
		AskStatus: string(answer.Status),
		AskAnswer: answeredOnly(answer),
	})

	waited := time.Since(pending.askedAt)
	if a.logger != nil {
		fields := []zap.Field{
			zap.String("session", a.session),
			zap.String("question", id),
			zap.String("status", string(answer.Status)),
			zap.Duration("waited", waited.Round(time.Millisecond)),
			zap.Int("options", len(q.Options)),
		}
		if answer.Answered() {
			a.logger.Info("chat: question answered", fields...)
		} else {
			// Not an error: it is exactly what the timeout exists for, and a
			// deployment that sees this often has a wait limit worth raising.
			a.logger.Info("chat: question not answered", fields...)
		}
	}
	return answer, nil
}

// answeredOnly returns the answer only when a person submitted one: a timeout has
// no answer to report, and an empty one would be indistinguishable from a person
// submitting nothing.
func answeredOnly(a tool.Answer) *tool.Answer {
	if !a.Answered() {
		return nil
	}
	return &a
}

// newQuestionID mints the handle a card's answer comes back with. It is random
// rather than sequential so a guess cannot consume somebody else's question; the
// route still checks it against the session it was asked in.
func newQuestionID() (string, error) {
	buf := make([]byte, 16)
	if _, err := rand.Read(buf); err != nil {
		return "", err
	}
	return hex.EncodeToString(buf), nil
}

// askTimeout is the configured wait limit, or the default.
func (s *Server) askTimeout() time.Duration {
	if s.chat.AskTimeout > 0 {
		return s.chat.AskTimeout
	}
	return DefaultAskTimeout
}

// handleAnswerQuestion delivers one card's answer to the turn waiting on it.
//
// It is a separate request from the streaming one by necessity: that response is
// already committed to the event stream, and a browser cannot write a second
// request body on it. The question id is the join between the two.
func (s *Server) handleAnswerQuestion(ctx context.Context, c *app.RequestContext) {
	sessionID := c.Param("id")
	questionID := c.Param("qid")

	var body struct {
		Selected []string `json:"selected"`
		Text     string   `json:"text"`
	}
	if err := c.BindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid JSON body"})
		return
	}
	selected := trimAll(body.Selected)
	text := strings.TrimSpace(body.Text)
	if len(selected) == 0 && text == "" {
		c.JSON(http.StatusBadRequest, map[string]string{"error": "请选择一个选项，或自己输入内容后再提交"})
		return
	}

	if s.questions == nil {
		c.JSON(http.StatusNotFound, map[string]string{"error": errQuestionGone.Error()})
		return
	}
	pending, ok := s.questions.peek(questionID, sessionID)
	if !ok {
		c.JSON(http.StatusNotFound, map[string]string{"error": questionGoneMessage})
		return
	}

	// Validate against what the card actually offered. The checks are here rather
	// than in the tool because this is the only place that knows both what was
	// asked and what came back — and the model must never be handed an "answer"
	// that no person could have chosen from the card.
	if text != "" && !pending.question.AllowCustom {
		c.JSON(http.StatusBadRequest, map[string]string{"error": "这个问题只能从给出的选项里选，不能自己输入"})
		return
	}
	if len(selected) > 1 && !pending.question.MultiSelect {
		c.JSON(http.StatusBadRequest, map[string]string{"error": "这个问题只能选一个选项"})
		return
	}
	if bad := unknownOption(selected, pending.question.Options); bad != "" {
		c.JSON(http.StatusBadRequest, map[string]string{
			"error": fmt.Sprintf("选项 %q 不在这个问题里，请从卡片上的选项中选择", bad),
		})
		return
	}

	answer := tool.Answer{Status: tool.AnswerAnswered, Selected: selected, Text: text}
	if err := s.questions.resolve(questionID, sessionID, answer); err != nil {
		c.JSON(http.StatusNotFound, map[string]string{"error": questionGoneMessage})
		return
	}
	if s.logger != nil {
		s.logger.Info("chat: answer accepted",
			zap.String("session", sessionID),
			zap.String("question", questionID),
			zap.Int("selected", len(selected)),
			zap.Bool("custom", text != ""),
		)
	}
	c.JSON(http.StatusOK, map[string]bool{"ok": true})
}

// questionGoneMessage is what a stale submission is told. It names the likely
// causes because the common one is not a bug: the model gave up waiting, and the
// answer the user just typed arrived after the turn moved on.
const questionGoneMessage = "这个问题已经不在等待中了（可能已提交、已超时，或本轮已经结束），请让模型重新提问"

// trimAll trims every entry and drops the empty ones, so a UI that submits an
// empty selection is refused by the "nothing chosen" check rather than accepted
// as a choice named "".
func trimAll(in []string) []string {
	if len(in) == 0 {
		return nil
	}
	out := make([]string, 0, len(in))
	for _, s := range in {
		if t := strings.TrimSpace(s); t != "" {
			out = append(out, t)
		}
	}
	if len(out) == 0 {
		return nil
	}
	return out
}

// unknownOption returns the first selected label the question never offered, or
// "" when every label is one of them.
func unknownOption(selected []string, options []tool.Option) string {
	if len(selected) == 0 {
		return ""
	}
	offered := make(map[string]struct{}, len(options))
	for _, o := range options {
		offered[o.Label] = struct{}{}
	}
	for _, s := range selected {
		if _, ok := offered[s]; !ok {
			return s
		}
	}
	return ""
}
